cannon/mipsevm/multithreaded/stats.go

```diff
@@ -0,0 +1,147 @@
+	// State
+	lastLLStepByThread    *lru.Cache[Word, uint64]
+	isWakeupTraversal     bool
+	activeThreadId        Word
+	lastActiveStepThread0 uint64
+	// Stats
+	rmwSuccessCount        uint64
+	rmwFailCount           uint64
+	maxStepsBetweenLLAndSC uint64
+	// Tracks RMW reservation invalidation due to reserved memory being accessed outside of the RMW sequence
+	reservationInvalidationCount uint64
+	forcedPreemptionCount        uint64
+	failedWakeupCount            uint64
+	idleStepCountThread0         uint64
+}
+
+func (s *statsTrackerImpl) populateDebugInfo(debugInfo *mipsevm.DebugInfo) {
+	debugInfo.RmwSuccessCount = s.rmwSuccessCount
+	debugInfo.RmwFailCount = s.rmwFailCount
+	debugInfo.MaxStepsBetweenLLAndSC = s.maxStepsBetweenLLAndSC
+	debugInfo.ReservationInvalidationCount = s.reservationInvalidationCount
+	debugInfo.ForcedPreemptionCount = s.forcedPreemptionCount
+	debugInfo.FailedWakeupCount = s.failedWakeupCount
+	debugInfo.IdleStepCountThread0 = s.idleStepCountThread0
+}
+
+func (s *statsTrackerImpl) trackLL(threadId Word, step uint64) {
+	s.lastLLStepByThread.Add(threadId, step)
+}
+
+func (s *statsTrackerImpl) trackSCSuccess(threadId Word, step uint64) {
+	s.rmwSuccessCount += 1
+	s.recordStepsBetweenLLAndSC(threadId, step)
+}
+
+func (s *statsTrackerImpl) trackSCFailure(threadId Word, step uint64) {
+	s.rmwFailCount += 1
+	s.recordStepsBetweenLLAndSC(threadId, step)
+}
+
+func (s *statsTrackerImpl) recordStepsBetweenLLAndSC(threadId Word, scStep uint64) {
+	// Track rmw steps if we have the last ll step in our cache
+	if llStep, ok := s.lastLLStepByThread.Get(threadId); ok {
+		diff := scStep - llStep
+		if diff > s.maxStepsBetweenLLAndSC {
+			s.maxStepsBetweenLLAndSC = diff
+		}
+		// Purge ll step since the RMW seq is now complete
+		s.lastLLStepByThread.Remove(threadId)
+	}
+}
+
+func (s *statsTrackerImpl) trackReservationInvalidation() {
+	s.reservationInvalidationCount += 1
+}
+
+func (s *statsTrackerImpl) trackForcedPreemption() {
+	s.forcedPreemptionCount += 1
+}
+
+func (s *statsTrackerImpl) trackWakeupTraversalStart() {
+	s.isWakeupTraversal = true
+}
+
+func (s *statsTrackerImpl) trackWakeup() {
+	s.isWakeupTraversal = false
+}
+
+func (s *statsTrackerImpl) trackWakeupFail() {
+	if s.isWakeupTraversal {
+		s.failedWakeupCount += 1
+	}
+	s.isWakeupTraversal = false
+}
+
+func (s *statsTrackerImpl) trackThreadActivated(tid Word, step uint64) {
+	if s.activeThreadId == Word(0) && tid != Word(0) {
+		// Thread 0 has been deactivated, start tracking to capture idle steps
+		s.lastActiveStepThread0 = step
+	} else if s.activeThreadId != Word(0) && tid == Word(0) {
+		// Thread 0 has been activated, record idle steps
+		idleSteps := step - s.lastActiveStepThread0
+		s.idleStepCountThread0 += idleSteps
+	}
+	s.activeThreadId = tid
+}
+
+func NewStatsTracker() StatsTracker {
+	return newStatsTracker(5)
+}
+
+func newStatsTracker(cacheSize int) StatsTracker {
+	llStepCache, err := lru.New[Word, uint64](cacheSize)
```

Comment:
Could we use a lighterweight lru? golang-lru uses locks for every operation and we could do away with that overhead. The VM is driven by a single goroutine so thread-safety isn't a concern for us. However, I'm a bit concerned about the performance hit from locking. That said, ll/sc operations don't occur often enough so this is probably a moot concern.
Worth considering a simpler or custom construction if available. Not a blocker if golang-lru is easiest for us.